### PR TITLE
Make the registration fallback page use the r0 register endpoint

### PR DESCRIPTION
**Pull request by babolivier**
Instead of the deprecated v1 endpoint, which doesn't exist anymore in Synapse anyway.

### CODE REVIEW

**Comment by babolivier** [comment]
Turns out the API the page was implementing wasn't using the same semantics as the r0 endpoint. I'm currently fixing that.

**Comment by babolivier** [comment]
Right, I've now made it compatible with the r0 endpoint. Two issues are still there though:

* its implementation of ReCaptcha is so old that the remote JS script it depends to get the captcha logic now 404s, I've opened https://github.com/matrix-org/synapse/issues/6386 to track the resolution of this point
* it doesn't know (nor support) `m.login.terms`, i've opened https://github.com/matrix-org/synapse/issues/6387 to track the resolution of this point

@bmarty ^ this might be of interest to you.

File: synapse/static/client/register/js/register.js [hunk 1]
```diff
@@ -1,67 +1,93 @@
...
-        submitPassword(user, pwd, response.session);
-    }).error(function(err) {
-        Recaptcha.reload();
-        errorFunc(err);
-    });
+    $.ajax({
+        url: matrixRegistration.endpoint,
+        method: "POST",
+        data: JSON.stringify(data),
+        complete: function(response) {
```
Comment (richvdh) [inline synapse/static/client/register/js/register.js]: As a general rule, I prefer the promise syntax (`$.ajax({...}).done( f )`) over the callback syntax (`$.ajax({success: f})`). I find it easier to read, and has a lower tendency to lead to lots of indentation. In fact, all the reasons promises were introduced to JS.

In this case, separate `done` and `error` handlers would be clearer than a combined one which then has to go and figure out if we had success or fail.

File: synapse/static/client/register/js/register.js [hunk 2]
```diff
@@ -1,67 +1,93 @@
...
+                    var stage = flow.stages[j];
+                    if ("m.login.recaptcha" === stage) {
+                        serverExpectsCaptcha = true;
+                        break;
+                    }
+                }
+            }
+            if (!serverExpectsCaptcha) {
+                console.log("This server does not require a captcha.");
+                return;
```
Comment (richvdh) [inline synapse/static/client/register/js/register.js]: ... we should still complete the other auth stages, and call the onRegistered callback though?

File: synapse/static/client/register/js/register.js [hunk 3]
```diff
@@ -1,67 +1,93 @@
...
-        errorFunc(err);
-    });
+    $.ajax({
+        url: matrixRegistration.endpoint,
+        method: "POST",
+        data: JSON.stringify(data),
+        complete: function(response) {
+            if (response.responseJSON.errcode) {
+                if (response.status != 401) {
+                    Recaptcha.reload();
```
Comment (richvdh) [inline synapse/static/client/register/js/register.js]: why is this the right thing to do?

File: synapse/static/client/register/js/register.js [hunk 4]
```diff
@@ -1,67 +1,93 @@
...
+                }
+                errorFunc(response);
+                return
+            }
+            console.log("Success -> "+JSON.stringify(response));
+            submitPassword(user, pwd);
```
Comment (richvdh) [inline synapse/static/client/register/js/register.js]: what is this bit for? we've already submitted the password above.

File: synapse/static/client/register/js/register.js [hunk 5]
```diff
@@ -1,67 +1,93 @@
...
-        var serverExpectsCaptcha = false;
-        for (var i=0; i<response.flows.length; i++) {
-            var flow = response.flows[i];
-            if ("m.login.recaptcha" === flow.type) {
-                serverExpectsCaptcha = true;
-                break;
+    $.ajax({
+        url: matrixRegistration.endpoint,
+        method: "POST",
+        data: '{}',
```
Comment (richvdh) [inline synapse/static/client/register/js/register.js]: I think this is supposed to start with a basic request with no UIA, so just the desired username and password.

```suggestion
        data: JSON.stringify({username: user, password: pwd}),
```

... it's actually valid for a server to return a 200 to that, if it doesn't need recaptcha or whatever.

**Comment by bmarty** [comment]
Any update on this PR?

**Comment by richvdh** [comment]
@bmarty do you have a usecase for this? as per https://github.com/matrix-org/matrix-doc/issues/2411#issuecomment-574778979, I'm wondering if we can just drop it from the spec.

**Comment by richvdh** [comment]
ok I think we should drop this PR, and instead remove this code as per #7676.

**Pull request closed without merging**